django/db/migrations/migration.py

```diff
@@ -108,20 +121,39 @@ def apply(self, project_state, schema_editor, collect_sql=False):
                 schema_editor.collected_sql.append("-- %s" % operation.describe())
                 schema_editor.collected_sql.append("--")
                 if not operation.reduces_to_sql:
-                    continue
-            # Save the state before the operation has run
-            old_state = project_state.clone()
-            operation.state_forwards(self.app_label, project_state)
+                    return (from_state, to_state)
+            operation.state_forwards(self.app_label, to_state)
             # Run the operation
             atomic_operation = operation.atomic or (self.atomic and operation.atomic is not False)
             if not schema_editor.atomic_migration and atomic_operation:
                 # Force a transaction on a non-transactional-DDL backend or an
                 # atomic operation inside a non-atomic migration.
                 with atomic(schema_editor.connection.alias):
-                    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
+                    operation.database_forwards(self.app_label, schema_editor, from_state, to_state)
             else:
                 # Normal behaviour
-                operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
+                operation.database_forwards(self.app_label, schema_editor, from_state, to_state)
+            return (from_state, to_state)
+
+        def _apply_operations(operations):
+            if not operations:
+                return
+
+            for operation in operations:
+                from_state, to_state = _apply_operation(operation)
+                injected_operations = emit_post_operation_signal(
+                    migration=self,
+                    operation=operation,
+                    from_state=from_state,
+                    to_state=to_state,
```

Comment:
I think it's a good safeguard, the states should only be used for "reading".

Reply:
@charettes Do you think it would be better to do the `.clone()` where `emit_post_migrate_signal` gets called (`db/migrations/migration.py`) or inside the `emit_post_migrate_signal` function itself, prior to calling `models.signals.post_migrate.send`?

Reply:
I went with the former, let me know what you think.